IOPool/Output/src/RootOutputFile.cc

```diff
@@ -682,70 +684,53 @@ namespace edm {
 
-    ProductProvenanceRetriever const* provRetriever = nullptr;
-    {
-      EventPrincipal const* eventPrincipal = dynamic_cast<EventPrincipal const*>(&principal);
-      if (eventPrincipal) {
-        provRetriever =eventPrincipal->productProvenanceRetrieverPtr();
-        assert(provRetriever);
-      }
-    }
-    
+    bool const fastCloning = (branchType == InEvent) && (whyNotFastClonable_ == FileBlock::CanFastClone);
+    std::set<StoredProductProvenance> provenanceToKeep;
+    //
     //If we are dropping some of the meta data we need to know
     // which BranchIDs were produced in this process because
     // we may be storing meta data for only those products
+    // We do this only for event products.
     std::set<BranchID> producedBranches;
-    if(om_->dropMetaData() != PoolOutputModule::DropNone) {
-      for (auto bd: principal.productRegistry().allBranchDescriptions()) {
-        if(bd->produced()) {
+    if(doProvenance && branchType == InEvent && om_->dropMetaData() != PoolOutputModule::DropNone) {
+      Service<ConstProductRegistry> preg;
+      for(auto bd : preg->allBranchDescriptions()) {
+        if(bd->produced() && bd->branchType() == InEvent) {
           producedBranches.insert(bd->branchID());
         }
       }
     }
 
-    // Loop over EDProduct branches, fill the provenance, and write the branch.
+    // Loop over EDProduct branches, possibly fill the provenance, and write the branch.
     for(auto const& item : items) {
 
       BranchID const& id = item.branchDescription_->branchID();
       branchesWithStoredHistory_.insert(id);
 
       bool produced = item.branchDescription_->produced();
-      bool keepProvenance = productProvenanceVecPtr != nullptr &&
-                            (om_->dropMetaData() == PoolOutputModule::DropNone ||
-                             om_->dropMetaData() == PoolOutputModule::DropDroppedPrior ||
-                            (om_->dropMetaData() == PoolOutputModule::DropPrior && produced));
-      bool getProd = (produced || !fastCloning ||
-         treePointers_[branchType]->uncloned(item.branchDescription_->branchName()));
+      bool getProd = (produced || !fastCloning || treePointers_[InEvent]->uncloned(item.branchDescription_->branchName()));
+      bool keepProvenance = doProvenance && (produced || keepProvenanceForPrior);
 
       WrapperBase const* product = nullptr;
       ProductProvenance const* productProvenance = nullptr;
+      BasicHandle result;
       if(getProd) {
-        OutputHandle const oh = principal.getForOutput(id, mcc);
-        product = oh.wrapper();
-        productProvenance = oh.productProvenance();
-      } else {
-        if(keepProvenance) {
-          productProvenance = provRetriever->branchIDToProvenance(id);
+        bool found = occurrence.getByToken(item.token_, item.branchDescription_->unwrappedTypeID(), result);
+        product = result.wrapper();
+        if(found && keepProvenance) {
```

Comment:
Couldn't the product not be found but the provenance still be there? This would be the case where we did not write out the product but kept the provenance in a previous step.